syntax = "proto3";

//package market_services;

service MarketService {
	rpc registerSeller(ClientDetails) returns (BooleanResponse) {}
	rpc sellItem(ItemDetails) returns (ItemOpResp) {}
	rpc updateItem(ItemDetails) returns (ItemOpResp) {}
	rpc deleteItem(ItemDetails) returns (ItemOpResp) {}
	rpc displaySellerItems(ClientDetails) returns (ItemsList) {}

	rpc searchItem(SearchItemReq) returns (ItemsList) {}
}

message ClientDetails {
	string address = 1;
	string unique_id = 2;
}

message BooleanResponse {
	bool status = 1;
}

// "optional" is not specified explicitely, but depends on the functions, producing and consuming the messages.
message ItemDetails {
	int32 item_id = 1;
	string product_name = 2;
	oneof category {
		bool electronics = 3;
		bool fashion = 4;
		bool others = 5;
	}
	int32 quantity = 6;
	string description = 7;
	float price_per_unit = 8;
	ClientDetails seller_details = 9;
	float rating = 10;
}

message ItemsList {
	repeated ItemDetails items = 1;
}
// Item operation Response
message ItemOpResp {
	bool status = 1;
	optional int32 item_id = 2;
}

message SearchItemReq {
	string product_name = 1;
	oneof category {
		bool electronics = 2;
                bool fashion = 3;
                bool others = 4;
		bool any = 5;
	}
}
